src/optimized_transaction.rs

```diff
@@ -0,0 +1,210 @@
+                Ok(None) => {
+                    sleep(interval).await;
+                    elapsed += interval;
+                }
+                Err(err) => return Err(HeliusError::ClientError(err)),
+            }
+        }
+    }
+
+    /// Builds and sends an optimized transaction, and handles its confirmation status
+    ///
+    /// # Arguments
+    /// * `config` - The configuration for the smart transaction, which includes the transaction's instructions, the user's keypair, whether preflight checks
+    /// should be skipped, and how many times to retry the transaction, if provided
+    ///
+    /// # Returns
+    /// The transaction signature, if successful
+    pub async fn send_smart_transaction(&self, config: SmartTransactionConfig<'_>) -> Result<Signature> {
+        let pubkey: Pubkey = config.from_keypair.pubkey();
+        let mut recent_blockhash: Hash = self.connection().get_latest_blockhash()?;
+
+        // Build the initial transaction and estimate the priority fee
+        let mut transaction: Transaction = Transaction::new_with_payer(&config.instructions, Some(&pubkey));
+        transaction.try_sign(&[config.from_keypair], recent_blockhash)?;
+
+        // Serialize the transaction
+        let serialized_transaction: Vec<u8> =
+            serialize(&transaction).map_err(|e: Box<ErrorKind>| HeliusError::InvalidInput(e.to_string()))?;
+
+        // Convert the serialized transaction to a Base64 string
+        let transaction_base64: String = STANDARD.encode(&serialized_transaction);
+
+        // Get the priority fee estimate based on the serialized transaction
+        let priority_fee_request: GetPriorityFeeEstimateRequest = GetPriorityFeeEstimateRequest {
+            transaction: Some(transaction_base64),
+            account_keys: None,
+            options: None,
+        };
+
+        let priority_fee_estimate: GetPriorityFeeEstimateResponse =
+            self.rpc().get_priority_fee_estimate(priority_fee_request).await?;
+
+        let priority_fee = priority_fee_estimate
+            .priority_fee_estimate
+            .ok_or(HeliusError::InvalidInput(
+                "Priority fee estimate not available".to_string(),
+            ))?
+            .to_bits();
+
+        // Add the compute unit price instruction with the estimated fee
+        let compute_budget_ix: Instruction = ComputeBudgetInstruction::set_compute_unit_price(priority_fee);
+        let mut final_instructions: Vec<Instruction> = vec![compute_budget_ix];
+        final_instructions.extend(config.instructions.clone());
+
+        // Get the optimal compute units
+        if let Some(units) = self
+            .get_compute_units(final_instructions.clone(), pubkey, vec![])
+            .await?
+        {
+            // Add some margin to the compute units
+            let compute_units_ix: Instruction =
+                ComputeBudgetInstruction::set_compute_unit_limit((units as f64 * 1.1).ceil() as u32);
```

Comment:
why `* 1.1`?

Reply:
This is some extra margin to the compute units to ensure the transaction doesn't fail. Should I reduce it from `* 1.1` to something smaller?

Reply:
Is this a common pattern? Why not just set the CU's we calculated?